sonolark/lib/kube/kube.go

```diff
@@ -731,3 +741,80 @@ func (m *kubePackage) kubeGet(ctx context.Context, r *apiResource, wait time.Dur
 
 	// not reachable
 }
+
+func (m *kubePackage) kubePortForwardTestFn(t *starlark.Thread, b *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var namespace, pod, testPath string
+	var localPort, podPort int
+	unpacked := []interface{}{
+		"namespace", &namespace,
+		"pod", &pod,
+		"local_port", &localPort,
+		"pod_port", &podPort,
+		"test_path", &testPath,
+	}
+	if err := starlark.UnpackArgs(b.Name(), args, kwargs, unpacked...); err != nil {
+		return starlark.False, fmt.Errorf("<%v>: %v", b.Name(), err)
+	}
+
+	// check if test path is already reachable before port forward
+	testUrl := fmt.Sprintf("http://localhost:%v%v", localPort, testPath)
+
+	req, err := http.NewRequest(http.MethodGet, testUrl, nil)
+	if err != nil {
+		return starlark.False, err
+	}
+
+	ctx := t.Local(addon.GoCtxKey).(context.Context)
+	_, err = m.httpClient.Do(req.WithContext(ctx))
+	if err == nil {
+		return starlark.False, fmt.Errorf("endpoint %v already in use", testUrl)
+	}
+
+	// port forward the pod
+	roundTripper, upgrader, err := spdy.RoundTripperFor(m.config)
+	if err != nil {
+		return starlark.False, err
+	}
+
+	portForwardPath := fmt.Sprintf("/api/v1/namespaces/%v/pods/%v/portforward", namespace, pod)
+
+	host := strings.Replace(m.Master, "https://", "", 1)
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: roundTripper}, http.MethodPost, &url.URL{Scheme: "https", Path: portForwardPath, Host: host})
+
+	var errorBuffer, outBuffer bytes.Buffer
+	errorWriter := bufio.NewWriter(&errorBuffer)
+	outWriter := bufio.NewWriter(&outBuffer)
+
+	var readyChannel chan struct{}
+	portForwardStopChannel := make(chan struct{})
+
+	portForward, err := portforward.New(dialer, []string{fmt.Sprintf("%d:%d", localPort, podPort)}, portForwardStopChannel, readyChannel, outWriter, errorWriter)
+	if err != nil {
+		return starlark.False, err
+	}
+
+	go func() {
+		err = portForward.ForwardPorts()
+		if err != nil {
+			log.Fatal(err)
+		}
+	}()
+	defer close(portForwardStopChannel)
+
+	// check if endpoint is accessible after short delay
+	time.Sleep(125000000)
+
+	for retries := 0; retries < 120; retries++ {
+		resp, err := m.httpClient.Do(req.WithContext(ctx))
+		if err != nil {
+			time.Sleep(125000000)
+			continue
+		}
+
+		if resp.StatusCode == 200 {
```

Comment:
You may want to be more permissive here. 40x and 50x are errors, but depending on the endpoint maybe you'll end up with 10x, 20x, or 30x.

Reply:
So you could check just that `StatusCode < 400`

Reply:
I think this is something that still need changed to be more flexible. There are lots of non-error status codes that an endpoint could return.

Reply:
I think you're diverging from the point a little bit here, the point of creating this function was to simply check if we can get a response from an app which has been port forwarded using the Kubernetes API. I actually think it may be better to not even check the response status code and just return if the error is nil.